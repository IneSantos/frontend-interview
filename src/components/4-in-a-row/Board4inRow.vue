<template>
  <div class="board-4inRow">
    <div class="grid">
      <div v-for="(column, column_index) in 7" :key="column" class="col" @click="playMove(column_index)"  :ref="`col-${column_index}`">
        <div v-for="(row, row_index) in 6" :key="row">
            <div class="cell">
                <span v-if="!isDot(board[row_index][column_index])">{{board[row_index][column_index]}}</span>
                <span v-if="isDot(board[row_index][column_index])" class="dot" :class="[getDotColor(board[row_index][column_index])]"></span>
            </div>
        </div>
      </div>
    </div>


    <div class="score">
      <Player name="Player 1" :score="player1" class="score-info"/>
      <StopWatch ref="stopwatch" class="score-info"/>
      <Player name="Player 2" :score="player2" class="score-info"/>
    </div>

    <div v-if="finished5rounds" class="status">
      <div v-if="stalemate" class="result stale">It's a draw!<br>Stalemate!</div>
      <div v-else class="result"><span class="winner">{{winner}}</span> wins!</div>
    </div>

  </div>
</template>

<script>
import '@/styles/4-in-a-row/board4inRow.scss'
import Player from '../tic-tac-toe/Player';
import StopWatch from '../tic-tac-toe/StopWatch';

export default {
  name: 'Board4inRow',
  components: {
      Player,
      StopWatch
  },
  data() { 
    return {
      board: [
            ['','','','','','',''],
            ['','','','','','',''],
            ['','','','','','',''],
            ['','','','','','',''],
            ['','','','','','',''],
            ['','','','','','',''],
      ],
      player: "r",
      player1: 0,
      player2: 0,
      finished: false,
      stalemate: false,
      watchStarted: false,
      finished5rounds: false,
      winner: ''
    }
  },
  computed: {
    numOfgames () {
      return this.$store.getters.getNumOfGames;
    }
  },
  watch: {
    numOfgames: function () {
      this.finished5rounds =  this.numOfgames === 5 || this.player1 === 3 || this.player2 === 3;
      this.winner = this.player1 > this.player2 ? "P1" : "P2";
      this.$store.dispatch('setGameHistory', this.winner);
      
      setTimeout( () => {
        this.scrollToStats();
      }, 1000); 
    }
  },
  methods: {
        playMove(col) {            
            if (this.finished) {
                return;
            }

            this.updateBoard(col);
            this.$forceUpdate();

            if(!this.watchStarted){
                this.$refs.stopwatch.start();
                this.watchStarted = true;
                this.$store.dispatch('setGameStart', true);
            }
            
            if (this.checkFourConnected(this.player)) {
                this.finished = true;
            }
            else if (this.checkStalemate()) {
                this.stalemate = true;
                this.finished = true;
            } 
            
            this.updateScore();
            this.nextPlayer();
        },
        nextPlayer: function() {
            this.player = this.player === "r" ? "g" : "r";
        },  
        checkStalemate: function() {
            return !this.finished && !this.board.flat().includes('');
        },  
        isDot: function(value) {
            return value !== '';
        },
        getDotColor: function (value) {
            return value === 'r' ? 'red' : 'green';
        },
        resetBoard: function() {
            this.board = [
                ['','','','','','',''],
                ['','','','','','',''],
                ['','','','','','',''],
                ['','','','','','',''],
                ['','','','','','',''],
                ['','','','','','',''],
            ];
            this.finished = false;
            this.stalemate = false;  
            this.watchStarted = false;
        },
        updateScore: function() {
            if(!this.finished) return;

            if(!this.stalemate) this.player === "r" ? ++this.player1 :  ++this.player2;

            this.$store.dispatch('setWinsPlayer1', this.player1);
            this.$store.dispatch('setWinsPlayer2', this.player2);
            this.$store.dispatch('incNumOfGames');

            this.$refs.stopwatch.stop();
            this.$refs.stopwatch.reset();

            setTimeout( () => {
                this.resetBoard();
            }, 1200); 
        },
        scrollToStats: function() {
            if(!this.finished5rounds) return; 

            let container = document.getElementById("statistics");
            container.scrollIntoView({ behavior: 'smooth', block: 'end', inline: 'nearest'});
        },
        updateBoard: function(col) {
            for (let i = 0; i < 6; i++) {
                if(this.board[5-i][col] === '') {
                    this.board[5-i][col] = this.player;
                    break;
                }
            }
        },
        checkFourConnected: function (player) {
                // horizontalCheck 
                for (let j = 0; j < this.board[0].length-3 ; j++ ){
                    for (let i = 0; i < this.board.length ; i++){
                        if (this.board[i][j] == player && this.board[i][j+1] == player && this.board[i][j+2] == player && this.board[i][j+3] == player){
                            return true;
                        }           
                    }
                }
                // verticalCheck
                for (let i = 0; i < this.board.length -3 ; i++ ){
                    for (let j = 0; j < this.board[0].length; j++){
                        if (this.board[i][j] == player && this.board[i+1][j] == player && this.board[i+2][j] == player && this.board[i+3][j] == player){
                            return true;
                        }           
                    }
                }
                // ascendingDiagonalCheck 
                for (let i=3; i < this.board.length; i++){
                    for (let j=0; j < this.board[0].length-3; j++){
                        if (this.board[i][j] == player && this.board[i-1][j+1] == player && this.board[i-2][j+2] == player && this.board[i-3][j+3] == player)
                            return true;
                    }
                }
                // descendingDiagonalCheck
                for (let i=3; i < this.board.length; i++){
                    for (let j=3; j < this.board[0].length; j++){
                        if (this.board[i][j] == player && this.board[i-1][j-1] == player && this.board[i-2][j-2] == player && this.board[i-3][j-3] == player)
                            return true;
                    }
                }
            
            return false;
        }
  }

}
</script>