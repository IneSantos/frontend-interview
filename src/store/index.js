export const store = {
    state: {
      hasgameStarted: false,
      numOfgames: 0,
      scorePlayer1: 0,
      scorePlayer2: 0,
      winsPlayer1: 0,
      winsPlayer2: 0,
      lossPlayer1: 0,
      lossPlayer2:0,
      gameHistory: ['', '', '' , '', '']
    },
    mutations: {
      SET_GAME_STARTED(state, hasgameStarted) {
        state.hasgameStarted = hasgameStarted
      },
      INC_NUM_GAMES(state) {
        state.numOfgames++
      },
      SET_SCORE_PLAYER_1(state, scorePlayer1) {
        state.scorePlayer1 = scorePlayer1
      },
      SET_SCORE_PLAYER_2(state, scorePlayer2) {
        state.scorePlayer2 = scorePlayer2
      },
      SET_WINS_PLAYER_1(state, winsPlayer1) {
        state.winsPlayer1 = winsPlayer1
      },
      SET_WINS_PLAYER_2(state, winsPlayer2) {
        state.winsPlayer2 = winsPlayer2
      },
      SET_LOSS_PLAYER_1(state, lossPlayer1) {
        state.lossPlayer1 = lossPlayer1
      },
      SET_LOSS_PLAYER_2(state, lossPlayer2) {
        state.lossPlayer2 = lossPlayer2
      },
      SET_GAME_HISTORY(state, player) {
        state.push(player);
      },
      RESET_GAME(state) {
        state.hasgameStarted = false;
      }
    },
    actions: {
      setGameStart(context, hasgameStarted){
        context.commit('SET_GAME_STARTED', hasgameStarted);
      },
      resetBoard(context){
        context.commit('RESET_BOARD')
      },
      incNumOfGames(context){
        context.commit('INC_NUM_GAMES');
      },
      setScorePlayer1(context, scorePlayer1){
        context.commit('SET_SCORE_PLAYER_1', scorePlayer1);
      },
      setScorePlayer2(context, scorePlayer2){
        context.commit('SET_SCORE_PLAYER_2', scorePlayer2);
      },
      setWinsPlayer1(context, winsPlayer1){
        context.commit('SET_WINS_PLAYER_1', winsPlayer1);
      },
      setWinsPlayer2(context, winsPlayer2){
        context.commit('SET_WINS_PLAYER_2', winsPlayer2);
      },
      setLossPlayer1(context, lossPlayer1) {
        context.commit('SET_LOSS_PLAYER_1', lossPlayer1);
      },
      setLossPlayer2(context, lossPlayer2) {
        context.commit('SET_LOSS_PLAYER_2', lossPlayer2);
      },
      setGameHistory(context, player){
        context.commit('SET_GAME_HISTORY', player);
      }
    },
    getters: {
      getHasGameStarted(state) {
        return state.hasgameStarted;
      },
      getNumOfGames(state) {
        return state.numOfgames;
      },
      getScorePlayer1(state) {
        return state.scorePlayer1;
      },
      getScorePlayer2(state) {
        return state.scorePlayer2;
      },
      getWinsPlayer1(state) {
        return state.winsPlayer1;
      },
      getWinsPlayer2(state) {
        return state.winsPlayer2;
      },
      getLossPlayer1(state) {
        return state.lossPlayer1;
      },
      getLossPlayer2(state) {
        return state.lossPlayer2;
      },
      getGameHistory(state) {
        return state.gameHistory;
      },
    }

}

