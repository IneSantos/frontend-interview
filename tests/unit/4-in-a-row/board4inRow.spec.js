import { mount, shallowMount, createLocalVue } from '@vue/test-utils'
import Vuex from 'vuex'
import Board4inRow from '@/components/4-in-a-row/Board4inRow';
import { storeMocked } from '../../__mocks__/store.js'

const localVue = createLocalVue();
localVue.use(Vuex);

const store = new Vuex.Store(storeMocked);

describe('Board.vue', () => {
  it('renders default', () => {
    const wrapper = shallowMount( Board4inRow, { store, localVue });
    expect(wrapper).toMatchSnapshot();
  })

  it('renders finished', async () => {
    const wrapper = mount( Board4inRow, { store, localVue });
    await wrapper.setData({ finished5rounds: true});

    expect(wrapper).toMatchSnapshot();
  })

  it('Should call playMove when cell is clicked', async () => {
    const mockMethod = jest.spyOn(Board4inRow.methods, 'playMove');
    const wrapper = mount( Board4inRow, { store, localVue });

    await wrapper.findComponent({ ref: "col-0" }).trigger('click');
    expect(mockMethod).toHaveBeenCalled();
  })

  it('Should call checkWinner when cell is clicked', async () => {
    const mockMethod = jest.spyOn(Board4inRow.methods, 'checkFourConnected');
    const wrapper = mount( Board4inRow, { store, localVue });

    await wrapper.findComponent({ ref: "col-0" }).trigger('click');
    expect(mockMethod).toHaveBeenCalled();
  })

  it('Should call checkStalemate when cell is clicked', async () => {
    const mockMethod = jest.spyOn(Board4inRow.methods, 'checkStalemate');
    const wrapper = mount( Board4inRow, { store, localVue });

    await wrapper.findComponent({ ref: "col-0" }).trigger('click');
    expect(mockMethod).toHaveBeenCalled();
  })

  it('Should call updateScore when cell is clicked', async () => {
    const mockMethod = jest.spyOn(Board4inRow.methods, 'updateScore');
    const wrapper = mount( Board4inRow, { store, localVue });

    await wrapper.findComponent({ ref: "col-0" }).trigger('click');
    expect(mockMethod).toHaveBeenCalled();
  })

  it('Should change player when cell is clicked', async () => {
    const wrapper = mount( Board4inRow, { store, localVue });

    await wrapper.findComponent({ ref: "col-0" }).trigger('click');
    expect(wrapper.vm.$data.player).toBe('g');
  })  

  it('Should increment score of player1 when game is finished', async () => {
    const wrapper = mount( Board4inRow, { store, localVue });

    await wrapper.findComponent({ ref: "col-0" }).trigger('click');
    await wrapper.findComponent({ ref: "col-1" }).trigger('click');
    await wrapper.findComponent({ ref: "col-0" }).trigger('click');
    await wrapper.findComponent({ ref: "col-1" }).trigger('click');
    await wrapper.findComponent({ ref: "col-0" }).trigger('click');
    await wrapper.findComponent({ ref: "col-1" }).trigger('click');
    await wrapper.findComponent({ ref: "col-0" }).trigger('click');
    
    expect(wrapper.vm.$data.finished).toBeTruthy();
    expect(wrapper.vm.$data.player1).toBe(1);
  })
})
