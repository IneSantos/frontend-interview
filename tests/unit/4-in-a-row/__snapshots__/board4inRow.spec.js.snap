// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Board.vue renders default 1`] = `
<div class="board-4inRow">
  <div class="grid">
    <div class="col">
      <div>
        <div class="cell"><span></span>
          <!---->
        </div>
      </div>
      <div>
        <div class="cell"><span></span>
          <!---->
        </div>
      </div>
      <div>
        <div class="cell"><span></span>
          <!---->
        </div>
      </div>
      <div>
        <div class="cell"><span></span>
          <!---->
        </div>
      </div>
      <div>
        <div class="cell"><span></span>
          <!---->
        </div>
      </div>
      <div>
        <div class="cell"><span></span>
          <!---->
        </div>
      </div>
    </div>
    <div class="col">
      <div>
        <div class="cell"><span></span>
          <!---->
        </div>
      </div>
      <div>
        <div class="cell"><span></span>
          <!---->
        </div>
      </div>
      <div>
        <div class="cell"><span></span>
          <!---->
        </div>
      </div>
      <div>
        <div class="cell"><span></span>
          <!---->
        </div>
      </div>
      <div>
        <div class="cell"><span></span>
          <!---->
        </div>
      </div>
      <div>
        <div class="cell"><span></span>
          <!---->
        </div>
      </div>
    </div>
    <div class="col">
      <div>
        <div class="cell"><span></span>
          <!---->
        </div>
      </div>
      <div>
        <div class="cell"><span></span>
          <!---->
        </div>
      </div>
      <div>
        <div class="cell"><span></span>
          <!---->
        </div>
      </div>
      <div>
        <div class="cell"><span></span>
          <!---->
        </div>
      </div>
      <div>
        <div class="cell"><span></span>
          <!---->
        </div>
      </div>
      <div>
        <div class="cell"><span></span>
          <!---->
        </div>
      </div>
    </div>
    <div class="col">
      <div>
        <div class="cell"><span></span>
          <!---->
        </div>
      </div>
      <div>
        <div class="cell"><span></span>
          <!---->
        </div>
      </div>
      <div>
        <div class="cell"><span></span>
          <!---->
        </div>
      </div>
      <div>
        <div class="cell"><span></span>
          <!---->
        </div>
      </div>
      <div>
        <div class="cell"><span></span>
          <!---->
        </div>
      </div>
      <div>
        <div class="cell"><span></span>
          <!---->
        </div>
      </div>
    </div>
    <div class="col">
      <div>
        <div class="cell"><span></span>
          <!---->
        </div>
      </div>
      <div>
        <div class="cell"><span></span>
          <!---->
        </div>
      </div>
      <div>
        <div class="cell"><span></span>
          <!---->
        </div>
      </div>
      <div>
        <div class="cell"><span></span>
          <!---->
        </div>
      </div>
      <div>
        <div class="cell"><span></span>
          <!---->
        </div>
      </div>
      <div>
        <div class="cell"><span></span>
          <!---->
        </div>
      </div>
    </div>
    <div class="col">
      <div>
        <div class="cell"><span></span>
          <!---->
        </div>
      </div>
      <div>
        <div class="cell"><span></span>
          <!---->
        </div>
      </div>
      <div>
        <div class="cell"><span></span>
          <!---->
        </div>
      </div>
      <div>
        <div class="cell"><span></span>
          <!---->
        </div>
      </div>
      <div>
        <div class="cell"><span></span>
          <!---->
        </div>
      </div>
      <div>
        <div class="cell"><span></span>
          <!---->
        </div>
      </div>
    </div>
    <div class="col">
      <div>
        <div class="cell"><span></span>
          <!---->
        </div>
      </div>
      <div>
        <div class="cell"><span></span>
          <!---->
        </div>
      </div>
      <div>
        <div class="cell"><span></span>
          <!---->
        </div>
      </div>
      <div>
        <div class="cell"><span></span>
          <!---->
        </div>
      </div>
      <div>
        <div class="cell"><span></span>
          <!---->
        </div>
      </div>
      <div>
        <div class="cell"><span></span>
          <!---->
        </div>
      </div>
    </div>
  </div>
  <div class="score">
    <player-stub name="Player 1" score="0" class="score-info"></player-stub>
    <stopwatch-stub class="score-info"></stopwatch-stub>
    <player-stub name="Player 2" score="0" class="score-info"></player-stub>
  </div>
  <!---->
</div>
`;

exports[`Board.vue renders finished 1`] = `
<div class="board-4inRow">
  <div class="grid">
    <div class="col">
      <div>
        <div class="cell"><span></span>
          <!---->
        </div>
      </div>
      <div>
        <div class="cell"><span></span>
          <!---->
        </div>
      </div>
      <div>
        <div class="cell"><span></span>
          <!---->
        </div>
      </div>
      <div>
        <div class="cell"><span></span>
          <!---->
        </div>
      </div>
      <div>
        <div class="cell"><span></span>
          <!---->
        </div>
      </div>
      <div>
        <div class="cell"><span></span>
          <!---->
        </div>
      </div>
    </div>
    <div class="col">
      <div>
        <div class="cell"><span></span>
          <!---->
        </div>
      </div>
      <div>
        <div class="cell"><span></span>
          <!---->
        </div>
      </div>
      <div>
        <div class="cell"><span></span>
          <!---->
        </div>
      </div>
      <div>
        <div class="cell"><span></span>
          <!---->
        </div>
      </div>
      <div>
        <div class="cell"><span></span>
          <!---->
        </div>
      </div>
      <div>
        <div class="cell"><span></span>
          <!---->
        </div>
      </div>
    </div>
    <div class="col">
      <div>
        <div class="cell"><span></span>
          <!---->
        </div>
      </div>
      <div>
        <div class="cell"><span></span>
          <!---->
        </div>
      </div>
      <div>
        <div class="cell"><span></span>
          <!---->
        </div>
      </div>
      <div>
        <div class="cell"><span></span>
          <!---->
        </div>
      </div>
      <div>
        <div class="cell"><span></span>
          <!---->
        </div>
      </div>
      <div>
        <div class="cell"><span></span>
          <!---->
        </div>
      </div>
    </div>
    <div class="col">
      <div>
        <div class="cell"><span></span>
          <!---->
        </div>
      </div>
      <div>
        <div class="cell"><span></span>
          <!---->
        </div>
      </div>
      <div>
        <div class="cell"><span></span>
          <!---->
        </div>
      </div>
      <div>
        <div class="cell"><span></span>
          <!---->
        </div>
      </div>
      <div>
        <div class="cell"><span></span>
          <!---->
        </div>
      </div>
      <div>
        <div class="cell"><span></span>
          <!---->
        </div>
      </div>
    </div>
    <div class="col">
      <div>
        <div class="cell"><span></span>
          <!---->
        </div>
      </div>
      <div>
        <div class="cell"><span></span>
          <!---->
        </div>
      </div>
      <div>
        <div class="cell"><span></span>
          <!---->
        </div>
      </div>
      <div>
        <div class="cell"><span></span>
          <!---->
        </div>
      </div>
      <div>
        <div class="cell"><span></span>
          <!---->
        </div>
      </div>
      <div>
        <div class="cell"><span></span>
          <!---->
        </div>
      </div>
    </div>
    <div class="col">
      <div>
        <div class="cell"><span></span>
          <!---->
        </div>
      </div>
      <div>
        <div class="cell"><span></span>
          <!---->
        </div>
      </div>
      <div>
        <div class="cell"><span></span>
          <!---->
        </div>
      </div>
      <div>
        <div class="cell"><span></span>
          <!---->
        </div>
      </div>
      <div>
        <div class="cell"><span></span>
          <!---->
        </div>
      </div>
      <div>
        <div class="cell"><span></span>
          <!---->
        </div>
      </div>
    </div>
    <div class="col">
      <div>
        <div class="cell"><span></span>
          <!---->
        </div>
      </div>
      <div>
        <div class="cell"><span></span>
          <!---->
        </div>
      </div>
      <div>
        <div class="cell"><span></span>
          <!---->
        </div>
      </div>
      <div>
        <div class="cell"><span></span>
          <!---->
        </div>
      </div>
      <div>
        <div class="cell"><span></span>
          <!---->
        </div>
      </div>
      <div>
        <div class="cell"><span></span>
          <!---->
        </div>
      </div>
    </div>
  </div>
  <div class="score">
    <div class="player score-info">
      <h1>Player 1</h1>
      <p>0</p>
    </div>
    <div class="score-info">
      <p>00:00:00</p>
    </div>
    <div class="player score-info">
      <h1>Player 2</h1>
      <p>0</p>
    </div>
  </div>
  <div class="status">
    <div class="result"><span class="winner"></span> wins!</div>
  </div>
</div>
`;
